src/imars3dv2/filters/tilt.py

```diff
@@ -0,0 +1,301 @@
+            img0 = arrays[low_idx]
+            img1 = arrays[high_idx]
+            jobs.append(e.submit(calculate_tilt, img0, img1, low_bound, high_bound))
+    tilts = np.array([job.result() for job in jobs])
+    # use the average of the found tilt angles
+    tilt = np.mean(tilts)
+    # step 3: apply the tilt correction
+    if abs(tilt) < cut_off_angle_deg:
+        return arrays
+    else:
+        return apply_tilt_correction(arrays, tilt, ncore)
+
+
+def apply_tilt_correction(
+    arrays: np.ndarray,
+    tilt: float,
+    ncore: int = -1,
+) -> np.ndarray:
+    """
+    Apply the tilt correction to the given array.
+    For a 2 deg tilted rotation axis, this function will rotate each image -2
+    deg so that the rotation axis is upright.
+
+    Parameters
+    ----------
+    @param arrays:
+        The array for tilt correction
+    @param tilt:
+        The rotation axis tilt angle in degrees
+    @param ncore:
+        Number of cores to use for parallel median filtering, default is -1, which means using all available cores.
+
+    Returns
+    -------
+    @return:
+        The tilt corrected array
+    """
+    # dimensionality check
+    if arrays.ndim == 2:
+        return rotate(arrays, -tilt, resize=False, preserve_range=True)
+    elif arrays.ndim == 3:
+
+        def tilt_correct_image(idx):
+            return rotate(arrays[idx], -tilt, resize=False, preserve_range=True)
+
+        # NOTE: have to switch to threadpool as we are using a nested function
+        ncore = mproc.mp.cpu_count() if ncore == -1 else int(ncore)
+        with cf.ThreadPoolExecutor(ncore) as e:
+            jobs = [e.submit(tilt_correct_image, idx) for idx in range(arrays.shape[0])]
+        arrays = np.array([job.result() for job in jobs])
+        return arrays
+    else:
+        raise ValueError("array must be a 2d/3d array")
+
+
+def find_180_deg_pairs_idx(
+    angles: np.ndarray,
+    atol: float = 1e-3,
+    in_degrees: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Return the indices of the 180 degree pairs from given list of angles.
+
+    Parameters
+    ----------
+    @param angles:
+        The list of angles as a 1d array.
+    @param atol:
+        The absolute tolerance in degree for the 180 degree pairs.
+    @param in_degrees:
+        Whether the angles are in degrees or radians, default is in degrees.
+
+    Returns
+    -------
+    @return:
+        The indices of the 180 degree pairs (low_range, high_range)
+    """
+    # ensure correct dimension
+    if angles.ndim != 1:
+        raise ValueError("angles must be a 1d array")
+    # ensure angles are in degrees
+    angles = angles if in_degrees else np.degrees(angles)
+    # compute the self difference matrix
+    angles = angles[..., np.newaxis]
+    diff_matrix = angles.T - angles
+    # find the indices of the 180 degree pairs
+    idx_lowrange, idx_highrange = np.where(np.isclose(diff_matrix, 180, atol=atol))
```

Comment:
What does this return when there are multiple matches? e.g. The user measures 0, 90, 180, 270

Reply:
here is a quick test with IPyhon
```python
In [2]: import numpy as np

In [3]: a = np.array([0, 90, 180, 270])

In [4]: a = a[..., np.newaxis]

In [5]: diff = a.T - a

In [6]: diff
Out[6]: 
array([[   0,   90,  180,  270],
       [ -90,    0,   90,  180],
       [-180,  -90,    0,   90],
       [-270, -180,  -90,    0]])

In [7]: np.where(np.isclose(diff, 180))
Out[7]: (array([0, 1]), array([2, 3]))
```